base/applications/mspaint/mouse.cpp

```diff
@@ -58,216 +62,446 @@ BOOL nearlyEqualPoints(INT x0, INT y0, INT x1, INT y1)
+        if (bLeftButton)
         {
-            selectionModel.CalculateBoundingBoxAndContents(hdc);
-            if (selectionModel.PtStackSize() > 1)
+            imageModel.ResetToPrevious();
+            if (selectionModel.IsSrcRectSizeNonzero())
             {
-                selectionModel.DrawBackgroundPoly(hdc, bg);
+                selectionModel.CalculateContents(m_hdc);
+                selectionModel.DrawBackgroundRect(m_hdc, m_bg);
                 imageModel.CopyPrevious();
 
-                selectionModel.DrawSelection(hdc);
+                selectionModel.DrawSelection(m_hdc);
 
                 placeSelWin();
                 selectionWindow.ShowWindow(SW_SHOW);
                 ForceRefreshSelectionContents();
             }
-            selectionModel.ResetPtStack();
-            break;
         }
-        case TOOL_RECTSEL:
-            imageModel.ResetToPrevious();
-            if (selectionModel.IsSrcRectSizeNonzero())
-            {
-                selectionModel.CalculateContents(hdc);
-                selectionModel.DrawBackgroundRect(hdc, bg);
-                imageModel.CopyPrevious();
+    }
+    virtual void OnCancelDraw()
+    {
+        imageModel.ResetToPrevious();
+        selectionModel.ResetPtStack();
+        ToolBase::OnCancelDraw();
+    }
+};
 
-                selectionModel.DrawSelection(hdc);
+// TOOL_RUBBER
+struct RubberTool : ToolBase
+{
+    RubberTool() : ToolBase(TOOL_RUBBER)
+    {
+    }
 
-                placeSelWin();
-                selectionWindow.ShowWindow(SW_SHOW);
-                ForceRefreshSelectionContents();
-            }
-            break;
-        case TOOL_TEXT:
+    void draw(BOOL bLeftButton, LONG x, LONG y)
+    {
+        if (bLeftButton)
+            Erase(m_hdc, x, y, x, y, m_bg, toolsModel.GetRubberRadius());
+        else
+            Replace(m_hdc, x, y, x, y, m_fg, m_bg, toolsModel.GetRubberRadius());
+    }
+    virtual void OnDown(BOOL bLeftButton, LONG x, LONG y, BOOL bDoubleClick)
+    {
+        ToolBase::OnDown(bLeftButton, x, y, bDoubleClick);
+        imageModel.CopyPrevious();
+        draw(bLeftButton, x, y);
+    }
+    virtual void OnMove(BOOL bLeftButton, LONG x, LONG y)
+    {
+        draw(bLeftButton, x, y);
+        ToolBase::OnMove(bLeftButton, x, y);
+    }
+    virtual void OnUp(BOOL bLeftButton, LONG x, LONG y)
+    {
+        ToolBase::OnUp(bLeftButton, x, y);
+        draw(bLeftButton, x, y);
+    }
+    virtual void OnCancelDraw()
+    {
+        ToolBase::OnCancelDraw();
+    }
+};
+
+// TOOL_FILL
+struct FillTool : ToolBase
+{
+    FillTool() : ToolBase(TOOL_FILL)
+    {
+    }
+
+    virtual void OnDown(BOOL bLeftButton, LONG x, LONG y, BOOL bDoubleClick)
+    {
+        ToolBase::OnDown(bLeftButton, x, y, bDoubleClick);
+        imageModel.CopyPrevious();
+        if (bLeftButton)
+            Fill(m_hdc, x, y, m_fg);
+        else
+            Fill(m_hdc, x, y, m_bg);
+    }
+    virtual void OnMove(BOOL bLeftButton, LONG x, LONG y)
```

Comment:
Not sure whether this function and the other two need to have an explicit implementation.... since without them you will still end up calling the corresponding helpers of the base class... which is exactly the same behaviour as what you've implemented there!

Reply:
Deleted.